compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -118,39 +118,126 @@ object Types extends TypeUtils {
      *         a call to `resetInst`. This means all caches that rely on `isProvisional`
      *         can likely end up returning stale results.
      */
-    def isProvisional(using Context): Boolean = mightBeProvisional && testProvisional
+    // def isProvisional(using Context): Boolean = mightBeProvisional && testProvisional
+    def isProvisional(using Context): Boolean = mightBeProvisional && !currentProvisionalState.isEmpty
 
-    private def testProvisional(using Context): Boolean =
+    type ProvisionalState = util.HashMap[Type, Type]
+
+    def currentProvisionalState(using Context): ProvisionalState =
+      val state: ProvisionalState = util.HashMap()
```

Comment:
Aren't we paying some allocation cost here for `!mightBeProvisional` types, each time they call `signature`, `denot`, etc?  I'm surprised it hasn't shown up as a cost, so either I don't see why it's not costly or we're not seeing the performance cost for some reason.